packages/fx-core/src/component/driver/deploy/azure/AzureZipDeployDriver.ts

```diff
@@ -0,0 +1,212 @@
+// Licensed under the MIT license.
+
+import { AzureDeployDriver } from "./azureDeployDriver";
+import {
+  AxiosZipDeployResult,
+  AzureUploadConfig,
+  DeployStepArgs,
+} from "../../interface/buildAndDeployArgs";
+import { AzureResourceInfo, DriverContext } from "../../interface/commonArgs";
+import { TokenCredential } from "@azure/core-auth";
+import { IProgressHandler, LogProvider, UserInteraction } from "@microsoft/teamsfx-api";
+import { getLocalizedMessage, ProgressMessages } from "../../../messages";
+import { DeployConstant } from "../../../constant/deployConstant";
+import { createHash } from "crypto";
+import { default as axios } from "axios";
+import { DeployExternalApiCallError } from "../../../error/deployError";
+import { HttpStatusCode } from "../../../constant/commonConstant";
+
+export class AzureZipDeployDriver extends AzureDeployDriver {
+  pattern =
+    /\/subscriptions\/([^\/]*)\/resourceGroups\/([^\/]*)\/providers\/Microsoft.Web\/sites\/([^\/]*)/i;
+  private readonly serviceName: string;
+  protected helpLink = "https://aka.ms/teamsfx-actions/azure-app-service-deploy";
+  protected summaries: string[];
+  protected summaryPrepare: string[];
+  protected zipBuffer: Buffer | undefined;
+
+  constructor(
+    args: unknown,
+    context: DriverContext,
+    serviceName: string,
+    summaries: string[],
+    summaryPrepare: string[]
+  ) {
+    super(args, context);
+    this.serviceName = serviceName;
+    this.summaries = summaries;
+    this.summaryPrepare = summaryPrepare;
+  }
+
+  async azureDeploy(
+    args: DeployStepArgs,
+    azureResource: AzureResourceInfo,
+    azureCredential: TokenCredential
+  ): Promise<void> {
+    await this.progressBar?.start();
+    const cost = await this.zipDeploy(args, azureResource, azureCredential);
+    await this.progressBar?.next(ProgressMessages.restartAzureService);
+    await this.restartFunctionApp(azureResource);
+    await this.progressBar?.end(true);
+    if (cost > DeployConstant.DEPLOY_OVER_TIME) {
+      await this.context.logProvider?.info(
+        getLocalizedMessage(
+          "driver.deploy.notice.deployAcceleration",
+          "https://aka.ms/teamsfx-config-run-from-package"
+        ).localized
+      );
+    }
+  }
+
+  protected prepare: (args: DeployStepArgs) => Promise<void> = async (args: DeployStepArgs) => {
+    await this.progressBar?.next(ProgressMessages.packingCode);
+    await this.packageToZip(args, this.context);
+  };
+
+  /**
+   * deploy to azure app service or azure function use zip deploy method
+   * @param args local file needed to be deployed
+   * @param azureResource azure resource info
+   * @param azureCredential azure user login credential
+   * @return the zip deploy time cost
+   * @protected
+   */
+  public async zipDeploy(
+    args: DeployStepArgs,
+    azureResource: AzureResourceInfo,
+    azureCredential: TokenCredential
+  ): Promise<number> {
+    await this.progressBar?.next(ProgressMessages.packingCode);
```

Comment:
This is not very well handled, because the prepare method actually has a return value, but its return value is not constant. For example, in the state of ZIPdeploy its return value is Buffer, while if there is Docker Deploy, its return value may be a string.